_posts/2016-10-26-jwt-artsy-journey.md

```diff
@@ -0,0 +1,46 @@
+---
+layout: post
+title: "JWT, Artsy Journey"
+date: 2016-10-26 11:03
+comments: true
+author: ashkan
+categories: [Auth, JWT, API]
+---
+
+At Artsy we currently have ~1730 client applications hitting our API and requesting authentication. On a successful authentication, we want to be able to provide basic user/application data in the token so clients can use it without needing to call our API. For that we want to use JWT.
+
+JWT (JSON Web Token) is a self-contained, secure way of transmitting data between applications and clients as JSON objects. Once decoded, JWT contains a JSON including 3 separate secions, header, payload and signature. Header section is used to define the type and details about the token itself. Payload seciton includes _claims_ and data provided by token issuer about the user/application, things like token expiration time, user id and etc. Signature section is used to verify the sender of the JWT and make sure token wasn't modified along the way.
+<!-- more -->
+
+# Where we were
+
+After a successful login, our API would generate our own access token. Our encrypted access token included some basic information about the user and the application using the token.
+For every authenticated request we would have to decrypt the token, make sure user and application are valid and application still has access to our API. We are already _stateless_ since we don't store tokens in our database.
+We can share access tokens between applications and then each application has to make a call to our API to get basic user data.
+
+
+# Where we are going
+We want to keep our current auth flow which is already stateless and mainly replace our in-house generated access token to a more standard JWT. What that gives us is:
+
+- Be able to decode and read basic information out of our token. With our in-house generated access token this could happen only if client application had duplicated our API side's decryption logic. With JWT each app can decode the token and get basic data out of it, it can check if token is expired or not and etc. If client is doing something that needs to make sure token is still valid, it can still call the API, like before, to make sure token is valid and get more data about it.
+
+- Possibly include different data in JWT payload for different applications, if a JWT token was created for the application that deals with conversations, it doesn't need to know about the user's auction related information.
```

Comment:
I was mainly thinking about how we can have different payload data per application. You think that's kind of given based on above?